File: ImperatorToCK3/Source/Mappers/RegionMapper/ImperatorRegionMapper.cpp
#include "ImperatorRegionMapper.h"
#include "Log.h"
#include "OSCompatibilityLayer.h"
#include <filesystem>
#include <fstream>
#include "ParserHelpers.h"

namespace fs = std::filesystem;

mappers::ImperatorRegionMapper::ImperatorRegionMapper(const std::string& imperatorPath)
{
	LOG(LogLevel::Info) << "-> Initializing Imperator Geography";
	
	auto areaFileName = imperatorPath + "/game/map_data/areas.txt";
	auto regionFileName = imperatorPath + "/game/map_data/regions.txt";
	
	std::ifstream areaStream(fs::u8path(areaFileName));
	if (!areaStream.is_open())
		throw std::runtime_error("Could not open game/map_data/areas.txt!");

	std::ifstream regionStream(fs::u8path(regionFileName));
	if (!regionStream.is_open())
		throw std::runtime_error("Could not open game/map_data/regions.txt!");
	
	loadRegions(areaStream, regionStream);

	areaStream.close();
	regionStream.close();
}

void mappers::ImperatorRegionMapper::loadRegions(std::istream& areaStream, std::istream& regionStream)
{
	registerAreaKeys();
	parseStream(areaStream);
	clearRegisteredKeywords();
	
	registerRegionKeys();
	parseStream(regionStream);
	clearRegisteredKeywords();

	linkRegions();
}



void mappers::ImperatorRegionMapper::registerRegionKeys()
{
	registerRegex(R"([\w_&]+)", [this](const std::string& regionName, std::istream& theStream) {
		const auto newRegion = std::make_shared<ImperatorRegion>(theStream);
		regions[regionName] = newRegion;
	});
	registerRegex(commonItems::catchallRegex, commonItems::ignoreItem);
}
void mappers::ImperatorRegionMapper::registerAreaKeys()
{
	registerRegex(R"([\w_&]+)", [this](const std::string& areaName, std::istream& theStream) {
		const auto newArea = std::make_shared<ImperatorArea>(theStream);
		areas[areaName] = newArea;
	});
	registerRegex(commonItems::catchallRegex, commonItems::ignoreItem);
}


bool mappers::ImperatorRegionMapper::provinceIsInRegion(const unsigned long long province, const std::string& regionName) const
{
	const auto& regionItr = regions.find(regionName);
	if (regionItr != regions.end() && regionItr->second)
		return regionItr->second->regionContainsProvince(province);

	// "Regions" are such a fluid term.
	const auto& areaItr = areas.find(regionName);
	if (areaItr != areas.end() && areaItr->second)
		return areaItr->second->areaContainsProvince(province);

	return false;
}

std::optional<std::string> mappers::ImperatorRegionMapper::getParentRegionName(const unsigned long long provinceID) const
{
	for (const auto& [regionName, region] : regions)
	{
		if (region && region->regionContainsProvince(provinceID))
		{
			return regionName;
		}
	}
	Log(LogLevel::Warning) << "Province ID " << provinceID << " has no parent region name!";
	return std::nullopt;
}


std::optional<std::string> mappers::ImperatorRegionMapper::getParentAreaName(const unsigned long long provinceID) const
{
	for (const auto& [areaName, area]: areas)
	{
		if (area && area->areaContainsProvince(provinceID))
			return areaName;
	}
	Log(LogLevel::Warning) << "Province ID " << provinceID << " has no parent area name!";
	return std::nullopt;
}



bool mappers::ImperatorRegionMapper::regionNameIsValid(const std::string& regionName) const
{
	if (regions.contains(regionName))
		return true;

	// Who knows what the mapper needs. All kinds of stuff.
	if (areas.contains(regionName))
		return true;

	return false;
}

void mappers::ImperatorRegionMapper::linkRegions()
{
	for (const auto& [regionName, region]: regions)
	{
		for (const auto& [requiredAreaName, requiredArea]: region->getAreas())
		{
			const auto& areaItr = areas.find(requiredAreaName);
			if (areaItr != areas.end())
			{
				region->linkArea(requiredAreaName, areaItr->second);
			}
			else
			{
				throw std::runtime_error("Region's " + regionName + " area " + requiredAreaName + " does not exist!");
			}
		}
	}
}

File: ImperatorToCK3/Mappers/Region/ImperatorRegionMapper.cs
using commonItems;
using commonItems.Collections;
using commonItems.Colors;
using commonItems.Mods;
using ImperatorToCK3.CommonUtils.Map;
using ImperatorToCK3.Imperator.Geography;
using System.Collections.Generic;

namespace ImperatorToCK3.Mappers.Region;

internal sealed class ImperatorRegionMapper(AreaCollection areaCollection, MapData irMapData) {
	public IdObjectCollection<string, ImperatorRegion> Regions { get; } = [];

	public void LoadRegions(ModFilesystem imperatorModFS, ColorFactory colorFactory) {
		Logger.Info("Initializing Imperator geography...");

		const string regionsFilePath = "map_data/regions.txt";
		Logger.Debug($"Imperator regions file location: {imperatorModFS.GetActualFileLocation(regionsFilePath)}");

		var parser = new Parser(implicitVariableHandling: true);
		RegisterRegionKeys(parser, colorFactory);
		parser.ParseGameFile(regionsFilePath, imperatorModFS);

		Logger.IncrementProgress();
	}

	private void RegisterRegionKeys(Parser parser, ColorFactory colorFactory) {
		parser.RegisterRegex(CommonRegexes.String, (reader, regionName) =>
			Regions.AddOrReplace(new ImperatorRegion(regionName, reader, areaCollection, colorFactory)));
		parser.IgnoreAndLogUnregisteredItems();
	}

	public bool ProvinceIsInRegion(ulong provinceId, string regionName) {
		if (Regions.TryGetValue(regionName, out var region)) {
			return region.ContainsProvince(provinceId);
		}
		// "Regions" are such a fluid term.
		return areaCollection.TryGetValue(regionName, out var area) && area.ContainsProvince(provinceId);
	}
	public bool RegionNameIsValid(string regionName) {
		// Who knows what the mapper needs. All kinds of stuff.
		return Regions.ContainsKey(regionName) || areaCollection.ContainsKey(regionName);
	}
	public string? GetParentRegionName(ulong provinceId) {
		_provinceToRegionIndex ??= BuildProvinceToRegionIndex();
		if (_provinceToRegionIndex.TryGetValue(provinceId, out var regionId)) {
			return regionId;
		}

		if (!irMapData.IsImpassable(provinceId)) {
			Logger.Warn($"I:R province ID {provinceId} has no parent region name!");
		}
		return null;
	}
	private Dictionary<ulong, string> BuildProvinceToRegionIndex() {
		var index = new Dictionary<ulong, string>();
		foreach (var region in Regions) {
			foreach (var area in region.Areas) {
				foreach (var provinceId in area.ProvinceIds) {
					index.TryAdd(provinceId, region.Id);
				}
			}
		}
		return index;
	}
	private Dictionary<ulong, string>? _provinceToRegionIndex;
	public string? GetParentAreaName(ulong provinceId) {
		foreach (var area in areaCollection) {
			if (area.ContainsProvince(provinceId)) {
				return area.Id;
			}
		}
		Logger.Warn($"I:R province ID {provinceId} has no parent area name!");
		return null;
	}
}